fuel-merkle/src/sparse/verify.rs

```diff
@@ -0,0 +1,341 @@
+use crate::{
+    common::{
+        path::{
+            Instruction,
+            Path,
+        },
+        Bytes32,
+        Prefix,
+    },
+    sparse::{
+        empty_sum,
+        proof::{
+            ExclusionProof,
+            InclusionProof,
+            Proof,
+        },
+        Node,
+    },
+};
+
+pub fn verify<K: Into<Bytes32>, V: AsRef<[u8]>>(key: K, value: &V, proof: Proof) -> bool {
+    match proof {
+        Proof::InclusionProof(proof) => verify_inclusion(key, value, proof),
+        Proof::ExclusionProof(proof) => {
+            value.as_ref() == empty_sum() && verify_exclusion(key, proof)
+        }
+    }
+}
+
+fn verify_inclusion<K: Into<Bytes32>, V: AsRef<[u8]>>(
+    key: K,
+    value: &V,
+    proof: InclusionProof,
+) -> bool {
+    let InclusionProof { root, proof_set } = proof;
+
+    let key: Bytes32 = key.into();
+    let leaf = Node::create_leaf(&key, value);
+    let path = leaf.leaf_key();
+    let mut current = *leaf.hash();
+
+    for (i, side_hash) in proof_set.iter().enumerate() {
+        let index = u32::try_from(proof_set.len() - 1 - i).expect("Index is valid");
+        let prefix = Prefix::Node;
+        current = match path.get_instruction(index).unwrap() {
+            Instruction::Left => Node::calculate_hash(&prefix, &current, side_hash),
+            Instruction::Right => Node::calculate_hash(&prefix, side_hash, &current),
+        };
+    }
+
+    current == root
+}
+
+fn verify_exclusion<K: Into<Bytes32>>(key: K, proof: ExclusionProof) -> bool {
+    let ExclusionProof {
+        root,
+        proof_set,
+        path,
+        hash,
+    } = proof;
+
+    if key.into() != path {
+        return false;
+    }
+
+    let mut current = hash;
```

Comment:
Hmmm, the first question is why we can't use `Node::Placeholder.hash()` here?

I assume because of the optimization that we have(like replacing the whole subtree with one leaf node). But then the next question is: What if `hash` is a hash of the actual leaf that is present in the tree? Then `ExclusionProof` will return `true` while it should return `false`. In other words, the malicious prover creates proofs of inclusion but converts it into proofs of exclusion by using `node.hash()` as a `hash` in the `ExclusionProof`

Reply:
Proof of exclusion is more than just proving that the key is a placeholder. If it was just proving that the key has a placeholder value, we could just check that the value was a placeholder, and we'd be done. And if it is a placeholder, we still need to know that the placeholder belongs to the tree.

You're correct: `hash` is not guaranteed to be the the placeholder value; it is defined by the value of the closest leaf to the requested key when generating the proof:
```
hash:  *actual_leaf.hash()
```
This is deliberate. The hash may or may not be a placeholder. While a placeholder is always an exclusion proof, a non-placeholder can be either an exclusion or inclusion proof. 

Why?

When generating a proof, we need a leaf that is part of the tree we for which we are generating the proof. Based on the current key-values of the tree, the 'actual leaf' (i.e., the closest leaf to the requested leaf during traversal) may be a placeholder or a non-placeholder. To later validate the proof, we need the hash of this leaf and the related proof set.

When validating an exclusion proof, we need to know that the 'actual leaf' (placeholder or non-placeholder) **belongs** to the tree. We are essentially doing a find operation on [disjoint sets](https://en.wikipedia.org/wiki/Disjoint-set_data_structure): We rebuild the root from the hash of the actual leaf and proof set, and compare roots to prove that this element belongs to this set. This also means that even placeholders can belong or not belong to the tree, which is why we need to be specific.

Reply:
Regarding the ability to produce erroneous proofs: It's possible we need to add additional logic to prevent someone from modifying a proof that still validates. I can add more tests to see if it is possible to forge a false positive proof.